numpy/core/getlimits.py

```diff
@@ -9,19 +9,198 @@
+                         negep=-53,
+                         minexp=-1022,
+                         maxexp=1024,
+                         it=52,
+                         iexp=11,
+                         ibeta=2,
+                         irnd=5,
+                         ngrd=0,
+                         eps=2.0 ** -52.0,
+                         epsneg=2.0 ** -53.0,
+                         huge=_get_next(_f64, numeric.inf, 0),
+                         tiny=2.0 ** -1022.0)
+
+# Known parameters for float80
+_ld = ntypes.longdouble
+_float80_ma = MachArLike(_ld,
+                         machep=-63,
+                         negep=-64,
+                         minexp=-16382,
+                         maxexp=16384,
+                         it=63,
+                         iexp=15,
+                         ibeta=2,
+                         irnd=5,
+                         ngrd=0,
+                         eps=exp2(_ld(-63)),
+                         epsneg=exp2(_ld(-64)),
+                         huge=_get_next(_ld, numeric.inf, 0),
+                         tiny=exp2(_ld(-16382)))
+
+# Guessed / known parameters for double double; see:
+# https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format#Double-double_arithmetic
+# These numbers have the same exponent range as float64, but extended number of
+# digits in the significand.
+_float_dd_ma = MachArLike(_ld,
+                          machep=-105,
+                          negep=-106,
+                          minexp=-1022,
+                          maxexp=1024,
+                          it=105,
+                          iexp=11,
+                          ibeta=2,
+                          irnd=5,
+                          ngrd=0,
+                          eps=exp2(_ld(-105)),
+                          epsneg=exp2(_ld(-106)),
+                          huge=_get_next(_ld, numeric.inf, 0),
+                          tiny=exp2(_ld(-1022)))
+
+
+# Key to identify the floating point type is tuple (number of bits in
+# significand (ignoring implicit first digit), maximum exponent, number of
+# bytes).
+_KNOWN_TYPES = {(52, 1024, 8): _float64_ma,
+                (23, 128, 4): _float32_ma,
+                (63, 16384, 16): _float80_ma,  # float80 on 64-bit
+                (63, 16384, 12): _float80_ma,  # float80 on 32-bit
+                (105, 1024, 16): _float_dd_ma,  # double double
+               }
+
+
+def _get_machar(dtype):
+    """ Get MachAr instance or MachAr-like instance """
+    params = _MACHAR_PARAMS.get(dtype)
+    if params is None:
+        raise ValueError(repr(dtype))
+    # Detect known / suspected types
+    digits = int(abs(round(log2(_get_eps(dtype)))))
+    huge = _get_next(dtype, numeric.inf, 0)
+    maxexp = None if isnan(huge) else int(round(log2(huge)))
+    bytes = numeric.dtype(dtype).itemsize
+    key = (digits, maxexp, bytes)
+    if key in _KNOWN_TYPES:
+        return _KNOWN_TYPES[key]
+    return _discover_machar(dtype)
```

Comment:
Maybe just raise an error? We only guarantee IEEE, and now the double double. There is also a Motorola extended precision, I don't know if we should worry about that.